psi4/driver/procrouting/dft_funcs/dict_builder.py

```diff
@@ -0,0 +1,186 @@
+#
+# @BEGIN LICENSE
+#
+# Psi4: an open-source quantum chemistry software package
+#
+# Copyright (c) 2007-2017 The Psi4 Developers.
+#
+# The copyrights for code used from other parties are included in
+# the corresponding files.
+#
+# This file is part of Psi4.
+#
+# Psi4 is free software; you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License as published by
+# the Free Software Foundation, version 3.
+#
+# Psi4 is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License along
+# with Psi4; if not, write to the Free Software Foundation, Inc.,
+# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+# @END LICENSE
+#
+"""
+Builder function
+"""
+from psi4 import core
+from psi4.driver.p4util.exceptions import *
+
+from . import dict_lda_funcs
+from . import dict_gga_funcs
+from . import dict_mgga_funcs
+from . import dict_hyb_funcs
+from . import dict_dh_funcs
+
+dict_functionals = {}
+dict_functionals.update(dict_lda_funcs.functional_list)
+dict_functionals.update(dict_gga_funcs.functional_list)
+dict_functionals.update(dict_mgga_funcs.functional_list)
+dict_functionals.update(dict_hyb_funcs.functional_list)
+dict_functionals.update(dict_dh_funcs.functional_list)
+
+def check_components(components):
+    # sanity checks definition of x_functionals and c_functionals:
+    # only 1 component can have "use_libxc": True
+    use_libxc = 0
+    for item in components.keys():
+        if "use_libxc" in components[item].keys() and components[item]["use_libxc"]:
+            use_libxc += 1
+    if use_libxc > 1:
+        return(False)
+    # require explicit X_HF or C_MP2 with defined alpha for complex definitions
+    #if len(components) > 1 and "X_HF" not in components.keys() and "C_MP2" not in components.keys():
+    #    return(False)
+    if "X_HF" in components.keys() and "alpha" not in components["X_HF"].keys():
+        return(False)
+    if "C_MP2" in components.keys() and "alpha" not in components["C_MP2"].keys():
+        return(False)
+    return(True)
+
+def build_superfunctional_from_dictionary(name, npoints, deriv, restricted):
```

Comment:
Once we settle on this, can we add a long doctoring to this to explain what the new format looks like.

Reply:
Did you have anything in particular in mind?